### PR TITLE
A bunch of CID fixes

### PR DESCRIPTION
**Pull request by jsquyres**
Fix a variety of low-impact Coverity issues.

### CODE REVIEW

File: opal/mca/btl/tcp/btl_tcp_proc.c [hunk 1]
```diff
@@ -452,6 +452,11 @@ int mca_btl_tcp_proc_insert( mca_btl_tcp_proc_t* btl_proc,
     max_peer_interfaces = proc_data->max_local_interfaces;
     peer_interfaces = (mca_btl_tcp_interface_t**)calloc( max_peer_interfaces, sizeof(mca_btl_tcp_interface_t*) );
     assert(NULL != peer_interfaces);
+    if (NULL == peer_interfaces) {
```
Comment (ggouaillardet) [inline opal/mca/btl/tcp/btl_tcp_proc.c]: shouldn't you remove the `assert()` ?

File: ompi/mca/topo/treematch/treematch/tm_tree.c [hunk 2]
```diff
@@ -132,8 +132,8 @@ void free_non_constraint_tree(tm_tree_t *tree)
 {
   int d = tree->dumb;
 
-  free_list_child(tree);
   free_tab_child(tree);
+  free_list_child(tree);
```
Comment (ggouaillardet) [inline ompi/mca/topo/treematch/treematch/tm_tree.c]: @bosilca can you please make sure this is pushed upstream ?

File: opal/mca/btl/tcp/btl_tcp_proc.c (hunk 1 above)
Comment (jsquyres) [reply opal/mca/btl/tcp/btl_tcp_proc.c]: I thought about removing the `assert()` and at the time, decided against it.  But it's probably 60/40 better to remove it and just rely on the returning of `OPAL_ERR_OUT_OF_RESOURCE`.  I'll update.

File: ompi/communicator/comm_cid.c [hunk 3]
```diff
@@ -1026,7 +1026,7 @@ static int ompi_comm_allreduce_group_broadcast (ompi_comm_request_t *request)
 {
     ompi_comm_allreduce_context_t *context = (ompi_comm_allreduce_context_t *) request->context;
     ompi_comm_cid_context_t *cid_context = context->cid_context;
-    ompi_request_t *subreq[2];
```
Comment (bosilca) [inline ompi/communicator/comm_cid.c]: This is not necessary as subreq is a temporary array and the requests inside are copied using the subreq_count. Moreover, from a technical perspective these are supposed to be set to MPI_REQUEST_NULL and not to 0. This comment apply everywhere where such array is used.

File: ompi/mca/fcoll/static/fcoll_static_file_write_all.c [hunk 4]
```diff
@@ -976,12 +975,12 @@ mca_fcoll_static_file_write_all (mca_io_ompio_file_t *fh,
...
+    if (NULL != local_iov_array){
+        free(local_iov_array);
+        local_iov_array = NULL;
+    }
 
-        if (NULL != local_iov_array){
-            free(local_iov_array);
-            local_iov_array = NULL;
-        }
+    if (my_aggregator == fh->f_rank) {
```
Comment (bosilca) [inline ompi/mca/fcoll/static/fcoll_static_file_write_all.c]: Reading through this function hurts ;)

File: ompi/mca/topo/treematch/treematch/tm_tree.c (hunk 2 above)
Comment (bosilca) [reply ompi/mca/topo/treematch/treematch/tm_tree.c]: Yep, I forwarded to the treematch maintainers.

File: opal/mca/btl/tcp/btl_tcp_component.c [hunk 5]
```diff
@@ -1183,8 +1183,8 @@ static int mca_btl_tcp_component_exchange(void)
                  if ((AF_INET == my_ss.ss_family) &&
                      (4 != mca_btl_tcp_component.tcp_disable_family)) {
                      memcpy(&addrs[current_addr].addr_inet,
-                             &((struct sockaddr_in*)&my_ss)->sin_addr,
-                             sizeof(addrs[0].addr_inet));
+                            &((struct sockaddr_in*)&my_ss)->sin_addr,
+                            sizeof(((struct sockaddr_in*)&my_ss)->sin_addr));
```
Comment (bosilca) [inline opal/mca/btl/tcp/btl_tcp_component.c]: How is that not the same ?

File: ompi/mca/fcoll/static/fcoll_static_file_write_all.c (hunk 4 above)
Comment (edgargabriel) [reply ompi/mca/fcoll/static/fcoll_static_file_write_all.c]: @bosilca don't worry too much about this function. The static component might be axed before the branch to 4.0

File: opal/mca/btl/tcp/btl_tcp_component.c (hunk 5 above)
Comment (jsquyres) [reply opal/mca/btl/tcp/btl_tcp_component.c]: sizeof(addrs[0].addr_inet) == 16.
sizeof(((struct sockaddr_in*)&my_ss)->sin_addr) == 4

Comment (bosilca) [reply opal/mca/btl/tcp/btl_tcp_component.c]: Right. How about sizeof(struct in_addr) to make out intentions clear.

Comment (jsquyres) [reply opal/mca/btl/tcp/btl_tcp_component.c]: I admit that I kinda like the style of `memcpy(dest, &src, sizeof(src))`.  Do you think that's not clear?

Comment (bosilca) [reply opal/mca/btl/tcp/btl_tcp_component.c]: In other places we use "sizeof(struct in_addr)", which I think is a good indicator that we are dealing with an IPv4 address.

As a special mention, the original code was correct, we are moving 16 bytes around, but only the sin_addr part is initialized.

Comment (jsquyres) [reply opal/mca/btl/tcp/btl_tcp_component.c]: Updated to `sizeof(struct in_addr)`.

File: ompi/communicator/comm_cid.c (hunk 3 above)
Comment (jsquyres) [reply ompi/communicator/comm_cid.c]: Ok, removed the 2 commits with these comm_cid fixes.

**Pull request merged**